### DISCUSSION TITLE
Peeking at window and opening it causes only background to show in glass

### DISCUSSION THREAD

**Discussion by OldManAlpha** [Maplespe/DWMBlurGlass#185] [Q&A]
**Describe the bug**

When hovering over a window thumbnail in the taskbar, waiting for windows to disappear, then opening the hovered thumbnail, the glass only shows the desktop. This gets corrected when a window behind is updated. One very strange thing about this is that **OBS does not show this bug**, despite seeing it on the physical screen. It does show up in screenshots, however.

**To Reproduce**
1. Open a window that does not constantly update
2. Hover over a thumbnail of another window in the taskbar until you can see the desktop, but hovering for too long seems to cause the issue to not appear
3. Click on the thumbnail

**Screenshots**

![Bug](https://github.com/Maplespe/DWMBlurGlass/assets/60587722/d38c26bc-6579-4310-b23b-a88cefaa7d3f)

This is how to reproduce it, despite it not showing up in OBS:

https://github.com/Maplespe/DWMBlurGlass/assets/60587722/783a5113-d491-42a5-8ef3-f538ae109e79

**Additional information (please complete the following information):**
 OS Version : Windows 10 21H2.19044.1288
 Software Version : 2.0.1

**What other third-party software do you use:**
UltraUXThemePatcher
OldNewExplorer
Windhawk (Aerexplorer; Remove UWP Titlebars; Taskbar Thumbnail Reorder)
StartIsBack++
Winaero Tweaker

**Additional context**

**Comment by Maplespe** [comment]
This seems to be an internal dwm **optimization** issue, dwm doesn't always update windows in real time to save resources, especially background windows, what you see is actually cached. In fact it has little to do with DWMBlurGlass, and we are not sure if we can optimize it. And it doesn't always happen on all systems, like the Windows 11 I'm using doesn't have this problem.

Since Win8 a compositing technique called [DirectComposition](https://learn.microsoft.com/en-us/windows/win32/directcomp/basic-concepts?source=recommendations) has been introduced, so the rendering principle is completely different from Win7, and until now, Microsoft has been updating and optimizing it.

Our `CustomBlur` method was created based on the dcomp technology and we can't interfere with the underlying compositing behavior, and I can't explain it to you particularly well.

Using `OldBlur` method may not have this problem, because OldBlur traverses the window tree to generate the background image at every frame, but this is also the reason why OldBlur is inefficient.

**Comment by OldManAlpha** [comment]
You're right, it appears OldBlur doesn't have this issue.

I don't know if it's relevant, but I find it fascinating how inconsistent this is with OBS. If I'm looking at the video preview while this bug occurs, 3 different things might happen:
1. The bug shows on the window, but not in OBS
![WindowNoOBS](https://github.com/Maplespe/DWMBlurGlass/assets/60587722/9bc51958-26e1-49b6-ac72-7af420ef2459)
2. The bug shows on both the window and in OBS
![OBSAndWindow](https://github.com/Maplespe/DWMBlurGlass/assets/60587722/bdecc196-50d4-422b-8297-db1e69f5dfa5)
3. The bug shows in OBS but not the window
![OBSNoWindow](https://github.com/Maplespe/DWMBlurGlass/assets/60587722/7ced4b49-3b19-400d-9468-b743f23c5b11)

**Comment by Maplespe** [comment]
This may have something to do with the underlying driver implementation of the Video Capture API, I'm not quite sure, because what the Video Capture API gets is not necessarily what the dwm renders, it should be something more underlying.
